runtime/src/main/java/org/corfudb/runtime/view/replication/ChainReplicationProtocol.java

```diff
@@ -61,6 +66,42 @@ public ILogData peek(Layout layout, long globalAddress) {
         return ret == null || ret.isEmpty() ? null : ret;
     }
 
+    /**
+     * Read data from a range created from the given addresses.
+     *
+     * This will be changed when an API is added to support multiple of ranges
+     *
+     * @param layout                The layout to use for the readAll.
+     * @param globalAddresses       A set of addresses to read from.
+     * @return                      A map of addresses to committed
+     *                              addresses, hole filling if necessary.
+     */
+    @Override
+    public Map<Long, ILogData> readAll(Layout layout, Set<Long> globalAddresses) {
+        Range<Long> range = Range.encloseAll(globalAddresses);
+        long startAddress = range.lowerEndpoint();
+        long endAddress = range.upperEndpoint();
+        int numUnits = layout.getSegmentLength(startAddress);
+        log.trace("ReadAll[{}-{}]: chain {}/{}", startAddress, endAddress, numUnits, numUnits);
+
+        Map<Long, LogData> logResult = CFUtils.getUninterruptibly(layout
+                .getLogUnitClient(startAddress, numUnits - 1)
+                .read(null, range)).getReadSet();
+
+        //in case of a hole, do a normal read and use its hole fill policy
```

Comment:
I would need the LogUnitClient.read(null, range) to return a TreeMap instead of a hashMap.

The change (hack ?) I have done on my side is the following in IReplicationProtocol.java:

```
-    default @Nonnull Map<Long, ILogData> readAll(Layout layout, Set<Long> globalAddresses) {
+    default @Nonnull TreeMap<Long, ILogData> readAll(Layout layout, Set<Long> globalAddresses) {
         return globalAddresses.parallelStream()
                 .map(a -> new AbstractMap.SimpleImmutableEntry<>(a, read(layout, a)))
-                .collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue()));
+                .collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue(),(v1, v2)-> v1, TreeMap::new));
```

Then I had to propagate the TreeMap everywhere. I am sure there is a more elegant way to do that.